Fix list editor panics during insertion

Frizi opened the pull request:
Fixes #6522 

- Fixed occasional panics when inserting or moving elements around.
- Improved mouse events handling in list editor to behave correctly during box selection.
- Removed vectorEditor feature flag, enabling the list editor permanently.
- Replaced node background shape with `Rectangle` and removed unnecessary separation between hover area and background. Switched node hover state to use new mouse events system.
- Adjusted size of hover area of the list editor such that it is harder do accidentally insert a new element and easier to drag the nodes around.
- Fixed mouse down event propagation inside list editor when clicking without dragging. This allows dropdowns nested within list items to work correctly.
![image](https://user-images.githubusercontent.com/919491/236062164-96c20652-d109-44db-9cb2-0cb89ce53809.png)
![image](https://user-images.githubusercontent.com/919491/236062101-5b35e1db-298c-43a7-8329-591eb9790439.png)
![image](https://user-images.githubusercontent.com/919491/236358329-b58d4ae5-6d45-4c02-a9b4-64a461cbbd4c.png)


### Important Notes

The mouse handling changes involve an unfortunate huge hack, where we enable mouse events on the mouse shape during box selection. That way we know for sure that no other shape will be able to receive mouse enter event. Then the list editor widget is modified to only actually respond to events when its background is hovered. We will definitely want a more proper way to handle mouse event contention, but it's definitely out of scope for current bugfixing.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

Frizi committed:
return correct index after inserting element to list editor with weak placeholders

Frizi committed:
make nodes with lists easier to drag, block list insertion during box select

Frizi committed:
finish removing vector editor feature flag

Frizi committed:
fix edit mode swith handling on top of list editor

Frizi committed:
cleanup

wdanilo commented on app/gui/view/graph-editor/src/component/node.rs:
```diff
@@ -728,17 +689,28 @@ impl Node {
+            let background_press = model.background.on_event::<mouse::Down>();
+            let input_press = model.input.on_event::<mouse::Down>();
+            input_as_background_press <- input_press.gate(&input.set_edit_ready_mode);
+            trace input_as_background_press;
```
trace - also check other places before merge, as its hard to catch all of them during review

wdanilo commented on lib/rust/ensogl/component/list-editor/src/lib.rs:
```diff
@@ -527,7 +531,7 @@ impl<T: display::Object + CloneRef + Debug> ListEditor<T> {
                 f!([model] (gap, gaps, pos, size, is_dragging, enable_all, enable_last) {
                     let is_close_x = pos.x > -gap && pos.x < size.x + gap;
-                    let is_close_y = pos.y > -gap && pos.y < size.y + gap;
+                    let is_close_y = pos.y > 0.0 && pos.y < size.y;
```
why this change? The gap was there to display `+` button when the mouse is cldse to the edge of lit editor. Without this change it will not be visible, right?

Frizi replied:
It is visible, the area is just intentionally made a little smaller. The gap was extending the active bounding box in both X and Y axis. Y axis extension wasn't desirable, as it made the insertions way too easy to do accidentally, especially when attempting to drag the node.

wdanilo replied:
Wait, but dragging the node by clicking above / below list editor should be hardly supported / unsupported. I mean, there is just a few pixels for that and assuming someone would be able to find these pixels is bad. I prefer the drag areas to be bigger to easier initiate dragging / insertion of elements. Of course, currently there is problem with node dragging because we don't have the icons on the left, but it is not solvable by making this are smaller IMO.

Frizi replied:
I will adjust that when working on new node design, as we will have slightly more area to work with. For now, it behaves much better that way. It is still not hard to drag the elements, you simply drag the text anywhere. Also, using `gap` to extend drag area vertically was a little surprising to me. I would prefer that to be a separate configuration input, if we need one.

wdanilo commented on lib/rust/ensogl/component/list-editor/src/lib.rs:
```diff
@@ -1077,8 +1082,9 @@ impl<T: display::Object + CloneRef + 'static> Model<T> {
                 index
             };
+            let index = self.item_or_placeholder_index_to_index(actual_index);
             self.reposition_items();
-            self.item_or_placeholder_index_to_index(actual_index)
+            index
```
Does this change does any logic change or it was just syntactic one?

Frizi replied:
It does change the logic. We have to reposition the items after calculating the item index. Repositioning will often delete collapsed placeholders, which made the index conversion incorrect and caused panics down the line.

wdanilo replied:
oh!

wdanilo commented on lib/rust/ensogl/core/src/display/object/instance.rs:
```diff
@@ -3012,6 +3012,14 @@ pub trait LayoutOps: Object {
         self.display_object().layout.margin.set(Vector2(margin, margin));
     }
 
+    /// Set vertical and horizontal margins of the object. Margin is the free space around the
+    /// object.
+    fn set_margin_vh(&self, vertical: impl Into<Unit>, horizontal: impl Into<Unit>) -> &Self {
+        let margin = Vector2(horizontal.into().into(), vertical.into().into());
+        self.display_object().layout.margin.set(margin);
+        self
+    }
```
why this function is needed? This object implements `display::Object` so you can directly do `.margin.set`? 
EDIT: oh, it sets it not on root but on layout - hmm, why? There should be a proper explanation what it does and why it exists if the whole object is a display object.

Frizi replied:
It's simply another convenience method to set margins using two values, one for verticals and one horizontals. We already have one to set all margins to given value, or to set all four independently. The vertical/horizontal one was missing from usual set you have in CSS.

wdanilo replied:
Please ignore my previous comment. I totally missed that this is addition to display objects. I was 100% sure this is in the list editor code!

Frizi committed:
Add proper event resumption handling and use it to support dropdowns inside list editors.

wdanilo commented on lib/rust/ensogl/core/src/display/object/event.rs:
```diff
@@ -18,12 +18,31 @@ use crate::display::object::instance::WeakInstance;
+    Cancelled(Phase),
+    // Event has been cancelled and [`InstanceDef::emit_event_impl`] function has returned.
+    Stopped(Phase),
+    Finished,
```
1. What is finished? It can be many things IMO, lack of dock makes it hard to understand.
2. If I understand the meaning correctly, I think the name `Stopped` should be refactored to `FinishedCancelled` to make it more obvious what is the distinction between the two last.

wdanilo commented on lib/rust/ensogl/core/src/display/object/event.rs:
```diff
@@ -49,11 +68,14 @@ pub struct SomeEvent {
+
+    fn from_event<T: 'static>(event: Event<T>) -> Self {
         let state = event.state.clone_ref();
         let current_target = event.current_target.clone_ref();
-        let captures = Rc::new(Cell::new(true));
-        let bubbles = Rc::new(Cell::new(true));
+        let captures = event.captures.clone_ref();
+        let bubbles = event.bubbles.clone_ref();
         Self { data: frp::AnyData::new(event), state, current_target, captures, bubbles }
     }
```
Why this is not From impl?

Frizi replied:
It needs to be private.

wdanilo commented on lib/rust/ensogl/core/src/display/object/instance.rs:
```diff
@@ -2538,40 +2536,58 @@ impl InstanceDef {
-                } else {
-                    break;
-                }
+    fn emit_event_impl(&self, event: &event::SomeEvent) {
```
This needs proper docs now. The code inside is nicely documented, but this becomes really complex and I'd be very thankful for some docs explaining everything it does in the way it does.

jdunkerley commented:
Quick feedback from trying this branch. The Vector Dropdowns now work :).

But has a detrimental affect on being able to connect links.

https://user-images.githubusercontent.com/4699705/236422595-8efca0c3-cab7-4548-8f44-cbd232ebd77d.mp4

jdunkerley commented:
Stuff from my trying to use this build today:
- As above dragging to establish a connection is hard.
- Ctrl clicking into edit a node, then moving the cursor by clicking doesn't seem to work.
- If a Vector is pointed at a non-vector (e.g. single value or a function) then you just get a dot.
<img width="156" alt="image" src="https://user-images.githubusercontent.com/4699705/236511899-d920d4e8-67a4-4221-9896-5ecb2be0a98f.png">
- Seemed to get extra dots on things that aren't vectors.
<img width="242" alt="image" src="https://user-images.githubusercontent.com/4699705/236511974-bd283321-f3ef-40cd-817a-2ca7cca84c9d.png">
- The + to add the first item doesn't seem to work.
- Comma's seems to be sometimes visible and sometimes not.
<img width="396" alt="image" src="https://user-images.githubusercontent.com/4699705/236512050-a76ad609-d243-4287-90d9-d534f380c153.png">

Frizi committed:
add more docs to event propagation

Frizi committed:
fix event dispatch with disabled bubbling

Frizi commented:
I have fixed the issues with edge dragging. My previous event resume implementation have broken the mouse down events in new event system. Now it works again and nodes can be connected as expected, including inside list editor. Ctrl clicking/edit mode is also fixed with that.

<img width="484" alt="image" src="https://user-images.githubusercontent.com/919491/236839809-81866696-cb83-4e80-8910-4391dc35078a.png">

vitvakatu commented:
# QA Report 🔴 

1. Adding new elements to the empty vector is not working – the grey circle is unclickable.

Also, `split` does not have the first argument inside the array, see the video:

https://github.com/enso-org/enso/assets/6566674/7034024d-e429-4e2f-8931-2b0521e5e858

2. When dragging the edge over the vector editor, the cursor behaves weirdly, changing the shape to plus when between placeholders. Clicking on a free space with a `plus` cursor uses the nearest placeholder, but removing the connection also removes the element from the vector. It is asymmetrical behavior, I think dropping a new element with a `plus` shape should add the element, not reuse the nearest one.
Moreover, it requires exceptionally gentle positioning of the mouse pointer – a slight movement of a few pixels to the side completely changes the target port, which can confuse users. ("Damn, I hate this pixel-hunting just to connect the edge to the element").


https://github.com/enso-org/enso/assets/6566674/5a98be15-53dc-4b28-b0ae-46ef40011bfa

Frizi committed:
Improved widget type selection, allowing for multiple derivations on same node without recursion.

Better logic for determining when to use list editor and type-based inference for its default value. Improved hover areas for element insertion points, removed button shape in empty list.

sylwiabr commented:
@Frizi what is the status? I would love to test it :)

Frizi committed:
improve widget identity handling to properly identify nodes with named arguments

Frizi committed:
fix deserialization for widget data with strings containing escape sequences

Frizi committed:
fmt

Frizi committed:
fmt

vitvakatu commented:
## QA Report 🔴 

In general, looks much nicer. I like how dropping the edge to the vector looks and works now. 

There are still issues that I spotted, but they might be fixed separately, I suppose:

1. On the node `table.rename_columns Map.from_vector`, the `Map.from_vector` part is not visible.

<img width="1512" alt="Screenshot 2023-05-15 at 20 39 41" src="https://github.com/enso-org/enso/assets/6566674/cbd59714-ceaa-40b5-b15e-c43f9db4e798">

2. Adding elements to the vector after `Map.from_vector` adds `Boolean.True`. That is probably incorrect (it's probably an engine issue, though).
3. Connecting the edge to this vector is hard. It's hard to find a free place. It seems like the free place is right on top of the square bracket. That seems weird. In the video, you can see that I try to target that point for a very long time. Again, pixel-hunting is depressing.

https://github.com/enso-org/enso/assets/6566674/13f35e60-4a3f-4833-bfa6-5a692b29f8f6

4. Sometimes (almost every time for me), dropping the edge there adds two elements.

https://github.com/enso-org/enso/assets/6566674/0d45c438-f79f-4967-8e14-45d8443441dc

5. After recording the previous video, I removed the first element of the vector, and the cursor got stuck in the `red cross` state. Hovering the node again fixed it, so it is not that important.

<img width="1512" alt="Screenshot 2023-05-15 at 20 43 49" src="https://github.com/enso-org/enso/assets/6566674/967c3fa3-94d9-4ae7-a4c7-14081fe0bcea">

6. There is a place where the cursor turns green on this node, and pressing LMB produces bizarre code modifications. I don't see the same behavior for other functions, like `String.split`.


https://github.com/enso-org/enso/assets/6566674/a46eb5b0-81de-4e94-aab4-0ab8ff381cbd

Frizi committed:
fix qa reported issues

Frizi committed:
make insertion point hover areas slightly wider

Frizi commented:
PR is now ready for another QA round.

Frizi committed:
revert handling any widget click as node selection

vitvakatu commented:
QA passed 💚

mergify[bot] committed:
Merge branch 'develop' into wip/frizi/list-editor-stabilization

Frizi committed:
raise wasm size limit
